dialects/regex/normalizers_regex.cpp

```diff
@@ -0,0 +1,258 @@
+            flatten_in_arg<ConjOrDisj>(seq_app->arg(), newArgs);
+        else
+            newArgs.push_back(proj);
+    }
+}
+
+template<class ConjOrDisj> std::vector<const Def*> flatten_in_arg(Ref arg) {
+    std::vector<const Def*> newArgs;
+    flatten_in_arg<ConjOrDisj>(arg, newArgs);
+    return newArgs;
+}
+
+template<class ConjOrDisj> Ref make_binary_tree(Ref type, DefArray args) {
+    assert(!args.empty());
+    auto& world = args.front()->world();
+    return std::accumulate(args.begin() + 1, args.end(), args.front(), [&type, &world](const Def* lhs, const Def* rhs) {
+        return world.raw_app(type, world.app(world.annex<ConjOrDisj>(), world.lit_nat(2)), world.tuple({lhs, rhs}));
+    });
+}
+
+Ref normalize_conj(Ref type, Ref callee, Ref arg) {
+    auto& world = type->world();
+    world.DLOG("conj {}:{} ({})", type, callee, arg);
+    if (arg->as_lit_arity() > 2) {
+        auto flatArgs = flatten_in_arg<conj>(arg);
+        return make_binary_tree<conj>(type, flatArgs);
+    }
+    if (arg->as_lit_arity() > 1) return world.raw_app(type, callee, arg);
+    return arg;
+}
+
+bool compare_re(const Def* lhs, const Def* rhs) {
+    auto lhs_range = thorin::match<range>(lhs);
+    auto rhs_range = thorin::match<range>(rhs);
+    if (lhs_range) {
+        // sort ranges by increasing lower bound
+        if (rhs_range) return Lit::as(lhs_range->arg()->proj(0)) < Lit::as(rhs_range->arg()->proj(0));
+        // ranges to the end
+        return false;
+    } else if (rhs_range) {
+        // ranges to the end
+        return true;
+    }
+    // keep
+    return true;
+}
+
+void make_vector_unique(std::vector<const Def*>& args) {
+    std::stable_sort(args.begin(), args.end(), &compare_re);
+    {
+        auto newEnd = std::unique(args.begin(), args.end());
+        args.erase(newEnd, args.end());
+    }
+}
+
+std::optional<std::pair<nat_t, nat_t>> merge_ranges(std::pair<nat_t, nat_t> a, std::pair<nat_t, nat_t> b) {
+    if (!(a.second + 1 < b.first || b.second + 1 < a.first)) {
+        return {
+            {std::min(a.first, b.first), std::max(a.second, b.second)}
+        };
+    }
+    return {};
+}
+
+bool is_in_range(std::pair<nat_t, nat_t> range, nat_t needle) {
+    return needle >= range.first && needle <= range.second;
+}
+
+auto get_range(const Def* rng) -> std::pair<nat_t, nat_t> {
+    auto rng_match = thorin::match<range, false>(rng);
+    return {Lit::as<std::uint8_t>(rng_match->arg(0)), Lit::as<std::uint8_t>(rng_match->arg(1))};
+}
+
+struct app_range {
+    World& w;
+    Ref operator()(std::pair<nat_t, nat_t> rng) {
+        return w.app(w.annex<range>(), w.tuple({w.lit_int(8, rng.first), w.lit_int(8, rng.second)}));
+    }
+};
+
+void merge_ranges(std::vector<const Def*>& args) {
+    auto rangesBegin = args.begin();
+    while (rangesBegin != args.end() && !thorin::match<range>(*rangesBegin)) rangesBegin++;
+    if (rangesBegin == args.end()) return;
+
+    std::set<const Def*> toRemove;
```

Comment:
we use `snake_case`. See [here](https://anydsl.github.io/thorin2/coding.html#autotoc_md1)

`to_remove`